code/modules/admin/data_input.dm

```diff
@@ -0,0 +1,217 @@
+				return
+			input = get_one_match(stub, /datum, use_concrete_types = FALSE, only_admin_spawnable = FALSE)
+
+		if (DATA_INPUT_COLOR)
+			input = input(custom_title || "Select color:", custom_message, default) as null|color
+
+		if (DATA_INPUT_TEXT)
+			input = input(custom_title || "Enter text:", custom_message, default) as null|message
+
+		if (DATA_INPUT_ICON)
+			input = input(custom_title || "Select icon:", custom_message) as null|icon
+
+		if (DATA_INPUT_BOOL)
+			//lines written by the utterly insane
+			input = alert(custom_title || "True or False?", custom_message + (!isnull(default) ? "(Default: [default ? "True" : "False"])" : null), "True", "False") == "True" ? TRUE : FALSE
+
+		if (DATA_INPUT_LIST)
+			//TODO uhhhhhhhh h
+
+		if (DATA_INPUT_FILE)
+			input = input(custom_title || "Select file:", custom_message) as null|file
+
+		if (DATA_INPUT_DIR)
+			input = input(custom_title || "Enter direction:", "Dir as text (e.g. North), case doesn't matter", default) as null|text
+			input = uppertext(input)
+			switch(input)
+				if("NORTH")
+					input = NORTH
+				if("SOUTH")
+					input = SOUTH
+				if("EAST")
+					input = EAST
+				if("WEST")
+					input = WEST
+				if("NORTHEAST")
+					input = NORTHEAST
+				if("SOUTHEAST")
+					input = SOUTHEAST
+				if("NORTHWEST")
+					input = NORTHWEST
+				if("SOUTHWEST")
+					input = SOUTHWEST
+				else
+					boutput(user, "<span class='alert>Invalid dir!</span>")
+					return
+
+		if (DATA_INPUT_JSON)
+			input = input(custom_title || "Enter JSON:", custom_message, json_encode(default)) as null|text
+			input = json_decode(input)
+
+		if (DATA_INPUT_REF)
+			input = input(custom_title || "Enter ref:", "brackets don't matter", null) as null|text
+			input = locate(input)
+			if (!input)
+				input = locate("\[[input]\]")
+			if (!input)
+				boutput(user, "<span class='alert'>Invalid ref.</span>")
+				return
+
+		if (DATA_INPUT_TURF_BY_COORDS)
+			var/x = input("X coordinate", "Set to turf at \[_, ?, ?\]", null) as null|num
+			var/y = input("Y coordinate", "Set to turf at \[[x], _, ?\]", null) as null|num
+			var/z = input("Z coordinate", "Set to turf at \[[x], [y], _\]", null) as null|num
+			input = locate(x, y, z)
+			if (!input)
+				boutput(user, "<span class='alert'>Invalid turf.</span>")
+				return
+
+		if (DATA_INPUT_REFPICKER)
+			var/datum/promise/promise = new
+			var/datum/targetable/refpicker/abil = new
+			abil.promise = promise
+			user.mob.targeting_ability = abil
+			user.mob.update_cursor()
+			input = promise.wait_for_value() //TODO timeout? maybe?
+
+		if (DATA_INPUT_NEW_INSTANCE)
+			var/stub = input(custom_title || "Enter part of type:", custom_message) as null|text
+			if (!stub)
+				boutput(user, "<span class='alert'>Cancelled.</span>")
+				return
+			input = get_one_match(stub, /datum, use_concrete_types = FALSE, only_admin_spawnable = FALSE)
+
+		if (DATA_INPUT_NUM_ADJUST)
+			input = input("Enter amount to adjust by:", custom_message) as null|num
```

Comment:
num adjust is meant to make relative adjustments to existing values, not act the same as num

Reply:
Right, but the input we want (a number) is the same as a num. The proc returns both the chosen data format and the data itself, so the caller can distinguish between a num and a num adjust.
Since a num adjust is pretty meaningless on its own (as a proc-call arg or whatever), I just let the caller decide what to do with it, like here for the VV menu-
https://github.com/goonstation/goonstation/blob/008ad337dba1db340febf8f7b3e6a9454a884e51/code/modules/admin/viewvariables/datumvars.dm#L835
https://github.com/goonstation/goonstation/blob/008ad337dba1db340febf8f7b3e6a9454a884e51/code/modules/admin/viewvariables/datumvars.dm#L858-L867